src/h3/_cy/memory.pyx

```diff
@@ -0,0 +1,115 @@
+from libc cimport stdlib
+from cython.view cimport array
+from .h3lib cimport H3int
+
+
+cdef size_t move_nonzeros(H3int* a, size_t n):
+    """ Move nonzero elements to front of array `a` of length `n`.
+    Return the number of nonzero elements.
+
+    Loop invariant: Everything *before* `i` or *after* `j` is "done".
+    Move `i` and `j` inwards until they equal, and exit.
+    You can move `i` forward until there's a zero in front of it.
+    You can move `j` backward until there's a nonzero to the left of it.
+    Anything to the right of `j` is "junk" that can be reallocated.
+
+    | a | b | 0 | c | d | ... |
+            ^           ^
+            i           j
+
+
+    | a | b | d | c | d | ... |
+            ^       ^
+            i       j
+    """
+    cdef:
+        size_t i = 0
+        size_t j = n
+
+    while i < j:
+        if a[j-1] == 0:
+            j -= 1
+            continue
+
+        if a[i] != 0:
+            i += 1
+            continue
+
+        # if we're here, we know:
+        # a[i] == 0
+        # a[j-1] != 0
+        # i < j
+        # so we can swap! (actually, move a[j-1] -> a[i])
+        a[i] = a[j-1]
+        j -= 1
+
+    return i
+
+
+cdef H3int[:] empty_memory_view():
+    # todo: get rid of this?
+    # there's gotta be a better way to do this...
+    # create an empty cython.view.array?
+    cdef:
+        H3int a[1]
+
+    return (<H3int[:]>a)[:0]
+
+
+cdef class H3MemoryManager:
```

Comment:
Please add a docstring as to what this class does. It's not immediately obvious it's specific to H3 indexes, although I think that's the case.